pymc_experimental/tests/custom_rjmcmc_sketch.py

```diff
@@ -0,0 +1,333 @@
+    return {i:np.vectorize(function_for_substance(i)) for i in range(len(initial_values))}
+
+
+class LogLike(tt.Op):
+    """
+    Specify what type of object will be passed and returned to the Op when it is
+    called. In our case we will be passing it a vector of values (the parameters
+    that define our model) and returning a single "scalar" value (the
+    log-likelihood)
+    """
+
+    itypes = [tt.dvector]  # expects a vector of parameter values when called
+    otypes = [tt.dscalar]  # outputs a single scalar value (the log likelihood)
+
+    def __init__(self, loglike):
+        """
+        Initialise the Op with various things that our log-likelihood function
+        requires. Below are the things that are needed in this particular
+        example.
+
+        Parameters
+        ----------
+        loglike:
+            The log-likelihood (or whatever) function we've defined
+        """
+
+        # add inputs as class attributes
+        self.loglikelihood = loglike
+        self.logpgrad = LogLikeGrad(self.loglikelihood)
+
+    def perform(self, node, inputs, outputs):
+        # the method that is used when calling the Op
+        (theta,) = inputs  # this will contain my variables
+
+        # call the log-likelihood function
+        logl = self.loglikelihood(theta)
+
+        outputs[0][0] = np.array(logl)  # output the log-likelihood
+
+    def grad(self, inputs, g):
+        # TODO test this somehow
+        (theta,) = inputs
+        return [g[0] * self.logpgrad(theta)]
+
+class LogLikeGrad(tt.Op):
+    """
+    TODO test this somehow
+    This Op will be called with a vector of values and also return a vector of
+    values - the gradients in each dimension.
+    """
+
+    itypes = [tt.dvector]
+    otypes = [tt.dvector]
+
+    def __init__(self, loglike):
+        """
+        Initialise with various things that the function requires. Below
+        are the things that are needed in this particular example.
+
+        Parameters
+        ----------
+        loglike:
+            The log-likelihood (or whatever) function we've defined
+        data:
+            The "observed" data that our log-likelihood function takes in
+        x:
+            The dependent variable (aka 'x') that our model requires
+        sigma:
+            The noise standard deviation that out function requires.
+        """
+
+        # add inputs as class attributes
+        self.likelihood = loglike
+
+    def perform(self, node, inputs, outputs):
+        (theta,) = inputs
+
+        outputs[0][0] = scipy.optimize.approx_fprime(theta, self.likelihood, epsilon= np.sqrt(np.finfo(float).eps))
+
+
+
+# Generate Data from the true model
+timepoints = np.linspace(0, 5, 1000)
+true_deltas = np.array([[0, 1],[0, 0]])
+true_ks     = np.array([[0, 2],[0, 0]])
+initial_values = np.array([5, 0])
+true_model = get_vectorized_solution_for_matrix(true_deltas,true_ks, initial_values)
+true_data = [true_model[i](timepoints) for i in range(2)]
```

Comment:
## General advice
ne shouldn't put execution code at the 0th indentation level.
That's because the Python interpreter will start executing this upon import, before even having parsed the entire file.
Also it leads to recursion problems with multiprocessing, because the child processes import the code again..

The typical pattern is this:
```
# Place this at the end of the file:
if __name__ == "__main__":
    # Now call other functions defined above:
    run_main_example()
```
The `__name__ == "__main__"` is only `True` if the script is primary entry point: `python my_script.py`

You can start adoping this pattern by simply indenting most of your code and putting it into a function.

## How you can make this into a test
If you also name the file to start with `test_` - for example `test_rjmcmc.py` - then `pytest` will recognize it as a file potentially containing tests.
Any top-level function named `test_*` will then be recognized as a test case.
For example: `def test_rjmcmc_example_one():`

Instead of `python custom_rjmcmc_sketch.py` you can then do [`pytest --verbose test_rjmcmc.py`](https://docs.pytest.org/en/7.0.x/how-to/usage.html).

...And that's how you can have your first test case.

Reply:
Consider it an integration test. You could run it with fewer iterations to make it faster.
A basic assert could be if it can sample to prior/posterior correctly.
You probably also have attributes on the step method instance that you could assert.
Also a seeded test that asserts individual posterior samples is an option.